src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/ClassResolver.java

```diff
@@ -0,0 +1,171 @@
+                try {
+                    ClassModel model = ClassFile.of().parse(jf.getInputStream(je).readAllBytes());
+                    ClassDesc desc =  model.thisClass().asSymbol();
+                    classMap.put(desc, new Info(m.moduleName(), m.path(), model));
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+        return new ScannedModuleClassResolver(loaded, classMap);
+    }
+
+    static ClassResolver forSystemModules(Runtime.Version version) {
+        String platformName = String.valueOf(version.feature());
+        PlatformProvider platformProvider = ServiceLoader.load(PlatformProvider.class).findFirst().orElseThrow();
+        PlatformDescription platform;
+        try {
+            platform = platformProvider.getPlatform(platformName, null);
+        } catch (PlatformProvider.PlatformNotSupported e) {
+            throw new JNativeScanFatalError("Release: " + platformName + " not supported", e);
+        }
+        JavaFileManager fm = platform.getFileManager();
+        return new SystemModuleClassResolver(fm);
+    }
+
+    record Info(String moduleName, Path jarPath, ClassModel model) {}
+
+    public abstract void forEach(BiConsumer<ClassDesc, ClassResolver.Info> action);
+    public abstract Optional<ClassResolver.Info> lookup(ClassDesc desc);
+
+    @Override
+    public abstract void close() throws IOException;
+
+    private static class ScannedModuleClassResolver extends ClassResolver {
+
+        private final List<JarFile> jars;
+        private final Map<ClassDesc, ClassResolver.Info> classMap;
+
+        public ScannedModuleClassResolver(List<JarFile> jars, Map<ClassDesc, Info> classMap) {
+            this.jars = jars;
+            this.classMap = classMap;
+        }
+
+        public void forEach(BiConsumer<ClassDesc, ClassResolver.Info> action) {
+            classMap.forEach(action);
+        }
+
+        public Optional<ClassResolver.Info> lookup(ClassDesc desc) {
+            return Optional.ofNullable(classMap.get(desc));
+        }
+
+        @Override
+        public void close() throws IOException {
+            for (JarFile jarFile : jars) {
+                jarFile.close();
+            }
+        }
+    }
+
+    private static class SystemModuleClassResolver extends ClassResolver {
+
+        private final JavaFileManager platformFileManager;
+        private final Map<String, String> packageToSystemModule;
+        private final Map<ClassDesc, Info> cache = new HashMap<>();
+
+        public SystemModuleClassResolver(JavaFileManager platformFileManager) {
+            this.platformFileManager = platformFileManager;
+            this.packageToSystemModule = packageToSystemModule();
+        }
+
+        private static Map<String, String> packageToSystemModule() {
+            List<ModuleDescriptor> descriptors = ModuleFinder.ofSystem()
```

Comment:
Is it a problem that we compute the package -> module map using the runtime info (so latest version) but then all the other info is taken from a release-specific symbol file? E.g. say that package "foo" was moved from module "A" to module "B" in version N, and that user passes N - 1 as release to the scan tool - would that work?

Reply:
That's a good point, I don't think that scenario will work. We should really use the release specific info if we can. I think that's relatively easy to do, will take a look.

Reply:
Ok, I managed to implement this, but I don't think we can actually test this use case, since (AFAIK) there's never been a case of a package being moved to a different module under the same name.